File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    hasCycle(head) {
        let pointer = head;
        const visited = new Set();

        while(pointer !== null){
            if(visited.has(pointer)){
                return true;
            }
            visited.add(pointer);
            pointer = pointer.next;
        }
        return false;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head or not head.next:
            return 


        slow , fast = head, head 

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev , curr = None , slow.next
        slow.next = None

        while curr:
            tmp = curr.next
            curr.next = prev 
            prev = curr
            curr = tmp

        first , second = head , prev

        while second:
            tmp1 = first.next
            tmp2 = second.next 

            first.next = second
            second.next = tmp1

            first = tmp1
            second = tmp2

        return 